README.md
# MicroPython Nextion library

[![Downloads](https://pepy.tech/badge/micropython-nextion)](https://pepy.tech/project/micropython-nextion)
![Release](https://img.shields.io/github/v/release/brainelectronics/micropython-nextion?include_prereleases&color=success)
![MicroPython](https://img.shields.io/badge/micropython-Ok-green.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

MicroPython Nextion library

---------------

## General

Ported library to interact with [Nextion serial displays][ref-nextion-wiki],
based on the [ITEAD Arduino Nextion][ref-itead-nextion-github] library.

## Installation

### Install required tools

Python3 must be installed on your system. Check the current Python version
with the following command

```bash
python --version
python3 --version
```

Depending on which command `Python 3.x.y` (with x.y as some numbers) is
returned, use that command to proceed.

```bash
python3 -m venv .venv
source .venv/bin/activate

pip install -r requirements.txt
```

## Setup

### Install package

Connect the MicroPython device to a network (if possible)

```python
import network
station = network.WLAN(network.STA_IF)
station.active(True)
station.connect('SSID', 'PASSWORD')
station.isconnected()
```

#### General

Install the latest package version of this lib on the MicroPython device

```python
import mip
mip.install("github:brainelectronics/micropython-nextion")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-nextion')
```

#### Specific version

Install a specific, fixed package version of this lib on the MicroPython device

```python
import mip
# install a verions of a specific branch
mip.install("github:brainelectronics/micropython-nextion", version="feature/add-mip-package-file")
# install a tag version
mip.install("github:brainelectronics/micropython-nextion", version="0.15.0")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-nextion')
```

#### Test version

Install a specific release candidate version uploaded to
[Test Python Package Index](https://test.pypi.org/) on every PR on the
MicroPython device. If no specific version is set, the latest stable version
will be used.

```python
import mip
mip.install("github:brainelectronics/micropython-nextion", version="0.15.0-rc1.dev29")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
# overwrite index_urls to only take artifacts from test.pypi.org
upip.index_urls = ['https://test.pypi.org/pypi']
upip.install('micropython-nextion')
```

See also [brainelectronics Test PyPi Server in Docker][ref-brainelectronics-test-pypiserver]
for a test PyPi server running on Docker.

### Manually

#### Upload files to board

Copy the module to the MicroPython board and import them as shown below
using [Remote MicroPython shell][ref-remote-upy-shell]

Open the remote shell with the following command. Additionally use `-b 115200`
in case no CP210x is used but a CH34x.

```bash
rshell --port /dev/tty.SLAB_USBtoUART --editor nano
```

Perform the following command inside the `rshell` to copy all files and
folders to the device

```bash
mkdir /pyboard/lib
mkdir /pyboard/lib/nextion

cp nextion/* /pyboard/lib/nextion

cp examples/basic/main.py /pyboard
cp examples/boot.py /pyboard
```

## Usage

Use one of the [examples](examples/) to get started. Read also the
[examples README](examples/README.md) to find all supported elements

## Credits

Big thank you to [ITEAD Studio][ref-itead-github] for the implementation
of the Arduino library.

<!-- Links -->
[ref-nextion-wiki]: https://wiki.iteadstudio.com/Nextion_HMI_Solution
[ref-itead-nextion-github]: https://github.com/itead/ITEADLIB_Arduino_Nextion
[ref-remote-upy-shell]: https://github.com/dhylands/rshell
[ref-brainelectronics-test-pypiserver]: https://github.com/brainelectronics/test-pypiserver
[ref-github-be-mircopython-modules]: https://github.com/brainelectronics/micropython-modules
[ref-itead-github]: https://github.com/itead


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [x.y.z] - yyyy-mm-dd
### Added
### Changed
### Removed
### Fixed
-->
<!--
RegEx for release version from file
r"^\#\# \[\d{1,}[.]\d{1,}[.]\d{1,}\] \- \d{4}\-\d{2}-\d{2}$"
-->

## Released
## [0.16.0] - 2023-06-12
### Added
- Validate `package.json` file with every test workflow run but without version validation
- Basic setup instructions added to docs examples file

### Changed
- Update `nextion/ulogging.py` to [version 0.5 of micropython-lib](https://github.com/micropython/micropython-lib/blob/7128d423c2e7c0309ac17a1e6ba873b909b24fcc/python-stdlib/logging/logging.py)

### Removed
- Outdated and unused `libs_external` folder with `ulogging.py`
- `libs_external` and `nextion/ulogging.py` removed from `.flake8` exclude list

### Fixed
- Installation instructions in Quickstart document are using `mip`

## [0.15.3] - 2023-05-17
### Fixed
- Publish releases to PyPi again as `micropython-nextion`, see #35

## [0.15.2] - 2023-05-14
### Fixed
- Logger import fixed, see #31
- Examples README improved with MicroPython port specific UART pin usage example, see #30

## [0.15.1] - 2023-05-14
### Fixed
- Remove yet unavailable files from `package.json`, see #33
- Add `ulogging` module to package to remove all external dependencies

## [0.15.0] - 2023-05-13
### Added
- `package.json` for `mip` installation with MicroPython v1.19.1 or newer, see #27
- Instructions for installation with `mip` on  MicroPython v1.19.1 or newer in `README`, see #27
- UART ID can be specified in `NexHardware` with `uart_id`, see #29

## [0.14.0] - 2023-03-01
### Added
- `.editorconfig` for common editor settings
- `.yamllint` to lint all used YAML files
- `requirements-test.txt` file
- Run YAML linter on test workflow
- `.readthedocs.yaml` definition file for ReadTheDocs
- `docs` folder containing example files and configurations, see #24
- `test-release` and `release` workflows create changelog based (pre-)releases
- Documentation for changing a value on a non active page, see #20

### Changed
- Removed unused files from `.gitignore` file
- Fail workflow on flake8 violations, see #23

### Fixed
- All workflows use checkout v3 instead of v2
- Fixed uncovered YAML syntax issues in all workflow files

## [0.13.0] - 2022-10-23
### Added
- `changelog2version` to [requirements file](requirements-deploy.txt)
- [GitHub CI test release workflow](.github/workflows/test-release.yml) to
  deploy package as `be-micropython-nextion` to
  [Test Python Package Index](https://test.pypi.org/) on every PR
- [Quickstart guide](QUICKSTART.md)

### Changed
- Use [`changelog2version`][ref-changelog2version-package] to create package
  version file, see [#19][ref-issue-19], in:
    - [GitHub CI test workflow](.github/workflows/test.yml)
    - [GitHub CI release workflow](.github/workflows/release.yml)

### Removed
- Script `update_version.py` to update version file

### Fixed
- Flake8 warning in [`nextion_waveform`](nextion/nextion_waveform.py) and
  [`nextion_hardware`](nextion/nextion_hardware.py), see [#17][ref-issue-17]
- Use `self._logger.*` instead of `self._nh._logger.*` in:
    - [`nextion_rtc`](nextion/nextion_rtc.py)
    - [`nextion_upload`](nextion/nextion_upload.py)
- Path of `main.py` to copy manually to the MicroPython board described in
  [`README`](README.md)

## [0.12.0] - 2022-07-30
### Added
- Support `NexRtc` usage with [`nextion_rtc`](nextion/nextion_rtc.py)

### Fixed
- Add missing documentation for [upload example](examples/upload/main.py) to
  [examples README](examples/)

## [0.11.0] - 2022-07-30
### Added
- Support `NexUpload` usage with [`nextion_upload`](nextion/nextion_upload.py)
- [Compiled example TFT file](examples/everything.tft) for
  [upload example](examples/upload/main.py)

## [0.10.0] - 2022-07-30
### Added
- Support `NexVariable` usage with
  [`nextion_variable`](nextion/nextion_variable.py)

### Fixed
- Add full line comment between different section of all examples

## [0.9.0] - 2022-07-30
### Added
- Support GPIO usage with [`nextion_gpio`](nextion/nextion_gpio.py)
- [Usage example](examples/gpio/main.py) of GPIO

## [0.8.0] - 2022-07-30
### Added
- Support all class specific functions of
    - Button
    - Dual state button
    - Number
    - Slider
    - Waveform

### Changed
- Add full line comment between different section of examples

### Fixed
- Flake8 warnings in several classes due to many imports and inheritance

## [0.7.1] - 2022-07-29
### Fixed
- Add class `CommonPointerMixin` to [`common`](nextion/common.py)

## [0.7.0] - 2022-07-29
### Added
- Support `Get_pointer_thickness_wid` and `Set_pointer_thickness_wid` by new
  class `CommonPointerMixin` in [`common`](nextion/common.py)

### Changed
- Inherit from `CommonPointerMixin` in
    - Gauge
    - Slider

## [0.6.0] - 2022-07-29
### Added
- Support `Get_place_xcen`, `Set_place_xcen`, `Get_place_ycen` and
  `Set_place_ycen` by new class `CommonPositionMixin` in
  [`common`](nextion/common.py)

### Changed
- Inherit from `CommonPositionMixin` in
    - Button
    - Dual state button
    - Number
    - Text

### Fixed
- Remove unused imports in nextion elements

## [0.5.0] - 2022-07-29
### Added
- Support `Get_background_color_bco` and `Set_background_color_bco` by new
  class `CommonBackgroundColorMixin` in [`common`](nextion/common.py)
- Support `Get_font_color_pco`, `Set_font_color_pco`, `getFont` and `setFont`
  by new class `CommonFontMixin` in [`common`](nextion/common.py)

### Changed
- Inherit from `CommonFontMixin` and `CommonBackgroundColorMixin` in
    - Button
    - Checkbox
    - Dual state button
    - Gauge
    - Number
    - Progressbar
    - Radio
    - Slider
    - Text
    - Waveform

### Fixed
- Remove unsupported `CommonValueMixin` from Button
- Some typos in comments


update_version.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
Update version info in python version file with latest changelog version

The changelog and package version shall always be aligned. This script can be
used to update the version info line "__version_info__ = ..." of a version.py
file with the latest changelog entry.
This changelog entry shall follow the semantic version pattern, see
https://semver.org/ and shall match the following pattern:

## [x.y.z] - yyyy-mm-dd

The line shall start with two hashtags followed by a single space. The semver
with x, y and z as non-negative integers, seperated by a dot and surrounded by
square brackets. Followed by a space, a dash, another space and the ISO8601
formatted date. Additional timestamps after the data, seperated from the date
by a single space, are optional.
"""

__author__ = "brainelectronics"
__copyright__ = "MIT License, Copyright (c) 2022 brainelectronics"
__credits__ = ["brainelectronics"]
__version__ = "0.1.0"
__maintainer__ = "brainelectronics"
__email__ = "git@brainelectronics.de"
__status__ = "Beta"

import argparse
import fileinput
import logging
from pathlib import Path
import re
import semver
from sys import stdout


def parser_valid_file(parser: argparse.ArgumentParser, arg: str) -> str:
    """
    Determine whether file exists.
    :param      parser:                 The parser
    :type       parser:                 parser object
    :param      arg:                    The file to check
    :type       arg:                    str
    :raise      argparse.ArgumentError: Argument is not a file
    :returns:   Input file path, parser error is thrown otherwise.
    :rtype:     str
    """
    if not Path(arg).is_file():
        parser.error("The file {} does not exist!".format(arg))
    else:
        return arg


def parse_arguments() -> argparse.Namespace:
    """
    Parse CLI arguments.
    :raise      argparse.ArgumentError  Argparse error
    :return:    argparse object
    """
    parser = argparse.ArgumentParser(description="""
    Update version info file based on changelog entry
    """, formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    # default arguments
    parser.add_argument('-d', '--debug',
                        action='store_true',
                        help='Output logger messages to stderr')

    # specific arguments
    parser.add_argument('--changelog_file',
                        dest='changelog_file',
                        required=True,
                        type=lambda x: parser_valid_file(parser, x),
                        help='Path to changelog file')

    parser.add_argument('--version_file',
                        dest='version_file',
                        required=True,
                        type=lambda x: parser_valid_file(parser, x),
                        help='Path to version file')

    parsed_args = parser.parse_args()

    return parsed_args


def parse_changelog(changelog_file: Path, logger: logging.Logger) -> str:
    """
    Parse the changelog for the first matching release version line

    :param      changelog_file:  The path to the changelog file
    :type       changelog_file:  Path
    :param      logger:          Logger object
    :type       logger:          logging.Logger

    :returns:   Extracted semantic version string
    :rtype:     str
    """
    release_version_line_regex = r"^\#\# \[\d{1,}[.]\d{1,}[.]\d{1,}\] \- \d{4}\-\d{2}-\d{2}"
    # append "$" to match only ISO8601 dates without additional timestamps
    release_version_line = ""

    with open(changelog_file, "r") as f:
        for line in f:
            match = re.search(release_version_line_regex, line)
            if match:
                release_version_line = match.group()
                break

    logger.debug("First matching release version line: '{}'".
                 format(release_version_line))

    return parse_semver_line(release_version_line, logger)


def parse_semver_line(release_version_line: str,
                      logger: logging.Logger) -> str:
    """
    Parse a release version line for a semantic version

    :param      release_version_line:  The release version line as described
    :type       release_version_line:  str
    :param      logger:                Logger object
    :type       logger:                logging.Logger

    :returns:   Semantic version string
    :rtype:     str
    """
    # release_version_line = "## [0.2.0] - 2022-05-19"
    semver_regex = r"\[\d{1,}[.]\d{1,}[.]\d{1,}\]"
    semver_string = "0.0.0"

    # extract semver from release version line
    match = re.search(semver_regex, release_version_line)
    if match:
        semver_string = match.group()
        # remove '[' and ']' from semver_string
        semver_string = re.sub(r"[\[\]]", "", semver_string)
        if not semver.VersionInfo.isvalid(semver_string):
            logger.error("Parsed SemVer string is invalid, check format")
            raise ValueError("Invalid SemVer string")
        logger.debug("Extracted SemVer string: '{}'".format(semver_string))
    else:
        logger.warning("No SemVer string found in given release version line")

    # semver_string = "0.2.0"
    return semver_string


def create_version_info_line(semver_string: str,
                             logger: logging.Logger) -> str:
    """
    Create the version info line used in "version.py"

    :param      semver_string:  The valid semver string
    :type       semver_string:  str
    :param      logger:         Logger object
    :type       logger:         logging.Logger

    :returns:   Complete version info line
    :rtype:     str
    """
    # semver_string = "0.2.0"
    ver = semver.VersionInfo.parse(semver_string)

    version_info = (str(ver.major), str(ver.minor), str(ver.patch))
    version_info_line = "__version_info__ = {}".format(version_info)
    logger.debug("New version info line: '{}'".format(version_info_line))

    # version_info_line = "__version_info__ = ('0', '2', '0')"
    return version_info_line


def update_version_file(version_file: Path,
                        version_info_line: str,
                        logger: logging.Logger) -> None:
    """
    Update the version file with a new version info line

    :param      version_file:       The path to the version file
    :type       version_file:       Path
    :param      version_info_line:  The version info line
    :type       version_info_line:  str
    :param      logger:             Logger object
    :type       logger:             logging.Logger
    """
    # standard output (print) is redirected to the original file
    for line in fileinput.input(version_file, inplace=True):
        if line.startswith("__version_info__ = "):
            print(version_info_line, end="\n")
        else:
            # keep line ending
            print(line, end="")

    logger.debug("Version file '{}' updated with '{}'".
                 format(version_file, version_info_line))


def main():
    # parse CLI arguments
    args = parse_arguments()

    custom_format = '[%(asctime)s] [%(levelname)-8s] [%(filename)-15s @'\
                    ' %(funcName)-15s:%(lineno)4s] %(message)s'
    logging.basicConfig(level=logging.INFO,
                        format=custom_format,
                        stream=stdout)
    logger = logging.getLogger(__name__)
    if args.debug:
        logger.setLevel(logging.DEBUG)

    here = Path(__file__).parent
    changelog_file = (here / args.changelog_file).resolve()
    version_file = (here / args.version_file).resolve()

    logger.debug("Using changelog file '{}' to update version file '{}'".
                 format(changelog_file, version_file))

    semver_string = parse_changelog(changelog_file, logger)
    version_info_line = create_version_info_line(semver_string, logger)
    update_version_file(version_file, version_info_line, logger)


if __name__ == '__main__':
    main()
